lib/AST/Attr.cpp

```diff
@@ -1438,6 +1515,45 @@ DifferentiatingAttr::create(ASTContext &context, bool implicit,
                                        std::move(original), linear, indices);
 }
 
+TransposingAttr::TransposingAttr(
+    ASTContext &context, bool implicit, SourceLoc atLoc, SourceRange baseRange,
+    DeclNameWithLoc original,
+    ArrayRef<ParsedAutoDiffParameter> params)
+    : DeclAttribute(DAK_Transposing, atLoc, baseRange, implicit),
+      Original(std::move(original)),
+      NumParsedParameters(params.size()) {
+  std::copy(params.begin(), params.end(),
+            getTrailingObjects<ParsedAutoDiffParameter>());
+}
+
+TransposingAttr::TransposingAttr(
+    ASTContext &context, bool implicit, SourceLoc atLoc, SourceRange baseRange,
+    DeclNameWithLoc original, AutoDiffIndexSubset *indices)
+    : DeclAttribute(DAK_Differentiating, atLoc, baseRange, implicit),
+    Original(std::move(original)), ParameterIndexSubset(indices) {}
+
+TransposingAttr *
+TransposingAttr::create(ASTContext &context, bool implicit,
+SourceLoc atLoc, SourceRange baseRange,
+DeclNameWithLoc original,
+ArrayRef<ParsedAutoDiffParameter> params) {
+  unsigned size = totalSizeToAlloc<ParsedAutoDiffParameter>(params.size());
+  void *mem = context.Allocate(size, alignof(DifferentiatingAttr));
+  return new (mem) TransposingAttr(context, implicit, atLoc, baseRange,
+                                   std::move(original), params);
+}
+
+TransposingAttr *
+TransposingAttr::create(ASTContext &context, bool implicit,
+SourceLoc atLoc, SourceRange baseRange,
+DeclNameWithLoc original,
+AutoDiffIndexSubset *indices) {
+  void *mem = context.Allocate(sizeof(DifferentiatingAttr),
```

Comment:
```suggestion
  void *mem = context.Allocate(sizeof(TransposingAttr),
```

Reply:
This is a fundamental issue that may be what's causing the undefined behavior you observed.

Reply:
Doesn't seem like this fixed the `llvm::nulls()` hack 😞